src/ring_theory/graded_algebra/basic.lean

```diff
@@ -158,3 +158,35 @@ begin
 end
 
 end graded_algebra
+
+
+section homogeneous_induction
+
+variables {ι R A: Type*} [decidable_eq ι]
+variables [comm_ring R] [semiring A] [algebra R A]
+variables (𝒜 : ι → submodule R A) (h : direct_sum.submodule_is_internal 𝒜)
+include h
+
+@[elab_as_eliminator]
+lemma set_like.homogeneous_induction {P : A → Prop}
+  (h_zero : P 0)
+  (h_hom : ∀ (a : A), set_like.is_homogeneous 𝒜 a → P a)
```

Comment:
Is this better or worse than
```suggestion
  (h_hom : ∀ {i} (a : 𝒜 i), P a)
```
Maybe it would help if you also made a PR showing how you intend to you use this, so we can decide what's most ueful.

Reply:
I used this in `Proj` construction; all the forms are usable.

Reply:
I will try just using `supr_induction` instead, so this pr might not be useful at all.

Reply:
Maybe the lemma that's worth having is `((h.supr_eq_top.ge : _) submodule.mem_top)`, as a shorthand for "in an internal direct sum, any element is a member of the supr"